# Segment & Cluster Toronto Neighbourhoods - KT

In this assignment, we will be required to explore, segment, and cluster the neighborhoods in the city of Toronto.

For the Toronto neighborhood data, a Wikipedia page exists that has all the information we need to explore and cluster the neighborhoods in Toronto. We will scrape the Wikipedia page and wrangle the data, clean it, and then read it into a pandas dataframe so that it is in a structured format.

Once the data is in a structured format, it will be ready to explore and cluster the neighborhoods in the city of Toronto.

The initial step is to import libraries for this project.

In [1]:
#Import Libraries
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Solving environment: done

# All requested packages already installed.

Solving environment: done

# All requested packages already installed.

Libraries imported.


## 1) Scrape the Wikipedia webpage

We will be scraping data from this website: https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M

#### Reference: https://www.youtube.com/watch?v=ng2o98k983k

In [2]:
# Install Beautiful Soup for Web-scraping
!easy_install beautifulsoup4
# Open up a new Terminal and run the following:
#pip install beautifulsoup4
print('BeautifulSoup4 is installed.')

Searching for beautifulsoup4
Best match: beautifulsoup4 4.8.0
Adding beautifulsoup4 4.8.0 to easy-install.pth file

Using /Users/KT/opt/anaconda3/lib/python3.7/site-packages
Processing dependencies for beautifulsoup4
Finished processing dependencies for beautifulsoup4
BeautifulSoup4 is installed.


In [3]:
from bs4 import BeautifulSoup
import requests

website_url = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text

soup = BeautifulSoup(website_url,'lxml')

Using Prettify, the website would be in a more presentable manner and easy to read!

In [4]:
print(soup.prettify())

<!DOCTYPE html>
<html class="client-nojs" dir="ltr" lang="en">
 <head>
  <meta charset="utf-8"/>
  <title>
   List of postal codes of Canada: M - Wikipedia
  </title>
  <script>
   document.documentElement.className="client-js";RLCONF={"wgBreakFrames":!1,"wgSeparatorTransformTable":["",""],"wgDigitTransformTable":["",""],"wgDefaultDateFormat":"dmy","wgMonthNames":["","January","February","March","April","May","June","July","August","September","October","November","December"],"wgMonthNamesShort":["","Jan","Feb","Mar","Apr","May","Jun","Jul","Aug","Sep","Oct","Nov","Dec"],"wgRequestId":"XeqbpgpAMFIAAIBEz90AAAAR","wgCSPNonce":!1,"wgCanonicalNamespace":"","wgCanonicalSpecialPageName":!1,"wgNamespaceNumber":0,"wgPageName":"List_of_postal_codes_of_Canada:_M","wgTitle":"List of postal codes of Canada: M","wgCurRevisionId":929562264,"wgRevisionId":929562264,"wgArticleId":539066,"wgIsArticle":!0,"wgIsRedirect":!1,"wgAction":"view","wgUserName":null,"wgUserGroups":["*"],"wgCategories":["Communi

Scrolling through the above result, we notice that the Postcode table is placed under **'table'**.
We can use *findAll()* in BeautifulSoup4 to only extract the **'table'**.

In [5]:
Wiki_table = soup.find_all('table')[0] 
df = pd.read_html(str(Wiki_table))
print(df[0].to_json(orient='records'))

[{"Postcode":"M1A","Borough":"Not assigned","Neighbourhood":"Not assigned"},{"Postcode":"M2A","Borough":"Not assigned","Neighbourhood":"Not assigned"},{"Postcode":"M3A","Borough":"North York","Neighbourhood":"Parkwoods"},{"Postcode":"M4A","Borough":"North York","Neighbourhood":"Victoria Village"},{"Postcode":"M5A","Borough":"Downtown Toronto","Neighbourhood":"Harbourfront"},{"Postcode":"M6A","Borough":"North York","Neighbourhood":"Lawrence Heights"},{"Postcode":"M6A","Borough":"North York","Neighbourhood":"Lawrence Manor"},{"Postcode":"M7A","Borough":"Queen's Park","Neighbourhood":"Not assigned"},{"Postcode":"M8A","Borough":"Not assigned","Neighbourhood":"Not assigned"},{"Postcode":"M9A","Borough":"Downtown Toronto","Neighbourhood":"Queen's Park"},{"Postcode":"M1B","Borough":"Scarborough","Neighbourhood":"Rouge"},{"Postcode":"M1B","Borough":"Scarborough","Neighbourhood":"Malvern"},{"Postcode":"M2B","Borough":"Not assigned","Neighbourhood":"Not assigned"},{"Postcode":"M3B","Borough":"No

In [14]:
Wiki_table = soup.find_all('table')[0] 

df_Toronto = pd.read_html(str(Wiki_table))[0]

Postcode = df_Toronto["Postcode"].tolist()
Borough = df_Toronto["Borough"].tolist()
Neighbourhood = df_Toronto["Neighbourhood"].tolist()

In [15]:
df_Toronto

Postcode           Borough  \
0        M1A      Not assigned   
1        M2A      Not assigned   
2        M3A        North York   
3        M4A        North York   
4        M5A  Downtown Toronto   
5        M6A        North York   
6        M6A        North York   
7        M7A      Queen's Park   
8        M8A      Not assigned   
9        M9A  Downtown Toronto   
10       M1B       Scarborough   
11       M1B       Scarborough   
12       M2B      Not assigned   
13       M3B        North York   
14       M4B         East York   
15       M4B         East York   
16       M5B  Downtown Toronto   
17       M5B  Downtown Toronto   
18       M6B        North York   
19       M7B      Not assigned   
20       M8B      Not assigned   
21       M9B         Etobicoke   
22       M9B         Etobicoke   
23       M9B         Etobicoke   
24       M9B         Etobicoke   
25       M9B         Etobicoke   
26       M1C       Scarborough   
27       M1C       Scarborough   
28       M1C       Scarborough   
29       M2C      Not assigned   
30       M3C        North York   
31       M3C        North York   
32       M4C         East York   
33       M5C  Downtown Toronto   
34       M6C              York   
35       M7C      Not assigned   
36       M8C      Not assigned   
37       M9C         Etobicoke   
38       M9C         Etobicoke   
39       M9C         Etobicoke   
40       M9C         Etobicoke   
41       M1E       Scarborough   
42       M1E       Scarborough   
43       M1E       Scarborough   
44       M2E      Not assigned   
45       M3E      Not assigned   
46       M4E      East Toronto   
47       M5E  Downtown Toronto   
48       M6E              York   
49       M7E      Not assigned   
50       M8E      Not assigned   
51       M9E      Not assigned   
52       M1G       Scarborough   
53       M2G      Not assigned   
54       M3G      Not assigned   
55       M4G         East York   
56       M5G  Downtown Toronto   
57       M6G  Downtown Toronto   
58       M7G      Not assigned   
59       M8G      Not assigned   
60       M9G      Not assigned   
61       M1H       Scarborough   
62       M2H        North York   
63       M3H        North York   
64       M3H        North York   
65       M3H        North York   
66       M4H         East York   
67       M5H  Downtown Toronto   
68       M5H  Downtown Toronto   
69       M5H  Downtown Toronto   
70       M6H      West Toronto   
71       M6H      West Toronto   
72       M7H      Not assigned   
73       M8H      Not assigned   
74       M9H      Not assigned   
75       M1J       Scarborough   
76       M2J        North York   
77       M2J        North York   
78       M2J        North York   
79       M3J        North York   
80       M3J        North York   
81       M4J         East York   
82       M5J  Downtown Toronto   
83       M5J  Downtown Toronto   
84       M5J  Downtown Toronto   
85       M6J      West Toronto   
86       M6J      West Toronto   
87       M7J      Not assigned   
88       M8J      Not assigned   
89       M9J      Not assigned   
90       M1K       Scarborough   
91       M1K       Scarborough   
92       M1K       Scarborough   
93       M2K        North York   
94       M3K        North York   
95       M3K        North York   
96       M4K      East Toronto   
97       M4K      East Toronto   
98       M5K  Downtown Toronto   
99       M5K  Downtown Toronto   
100      M6K      West Toronto   
101      M6K      West Toronto   
102      M6K      West Toronto   
103      M7K      Not assigned   
104      M8K      Not assigned   
105      M9K      Not assigned   
106      M1L       Scarborough   
107      M1L       Scarborough   
108      M1L       Scarborough   
109      M2L        North York   
110      M2L        North York   
111      M3L        North York   
112      M4L      East Toronto   
113      M4L      East Toronto   
114      M5L  Downtown Toronto   
115      M5L  Downtown Toronto   
116      M6L        North 

In [16]:
df_Toronto.shape

(287, 3)

There are 3 columns and 287 rows. However, there is some data cleaning to be done in this table. 

## 2) Data Cleaning

For instance, we will make the assumption of excluding all Postcodes that are *'Not assigned'* to a Borough.

In [17]:
df_Toronto=df_Toronto[df_Toronto.Borough != 'Not assigned']
df_Toronto

Postcode           Borough  \
2        M3A        North York   
3        M4A        North York   
4        M5A  Downtown Toronto   
5        M6A        North York   
6        M6A        North York   
7        M7A      Queen's Park   
9        M9A  Downtown Toronto   
10       M1B       Scarborough   
11       M1B       Scarborough   
13       M3B        North York   
14       M4B         East York   
15       M4B         East York   
16       M5B  Downtown Toronto   
17       M5B  Downtown Toronto   
18       M6B        North York   
21       M9B         Etobicoke   
22       M9B         Etobicoke   
23       M9B         Etobicoke   
24       M9B         Etobicoke   
25       M9B         Etobicoke   
26       M1C       Scarborough   
27       M1C       Scarborough   
28       M1C       Scarborough   
30       M3C        North York   
31       M3C        North York   
32       M4C         East York   
33       M5C  Downtown Toronto   
34       M6C              York   
37       M9C         Etobicoke   
38       M9C         Etobicoke   
39       M9C         Etobicoke   
40       M9C         Etobicoke   
41       M1E       Scarborough   
42       M1E       Scarborough   
43       M1E       Scarborough   
46       M4E      East Toronto   
47       M5E  Downtown Toronto   
48       M6E              York   
52       M1G       Scarborough   
55       M4G         East York   
56       M5G  Downtown Toronto   
57       M6G  Downtown Toronto   
61       M1H       Scarborough   
62       M2H        North York   
63       M3H        North York   
64       M3H        North York   
65       M3H        North York   
66       M4H         East York   
67       M5H  Downtown Toronto   
68       M5H  Downtown Toronto   
69       M5H  Downtown Toronto   
70       M6H      West Toronto   
71       M6H      West Toronto   
75       M1J       Scarborough   
76       M2J        North York   
77       M2J        North York   
78       M2J        North York   
79       M3J        North York   
80       M3J        North York   
81       M4J         East York   
82       M5J  Downtown Toronto   
83       M5J  Downtown Toronto   
84       M5J  Downtown Toronto   
85       M6J      West Toronto   
86       M6J      West Toronto   
90       M1K       Scarborough   
91       M1K       Scarborough   
92       M1K       Scarborough   
93       M2K        North York   
94       M3K        North York   
95       M3K        North York   
96       M4K      East Toronto   
97       M4K      East Toronto   
98       M5K  Downtown Toronto   
99       M5K  Downtown Toronto   
100      M6K      West Toronto   
101      M6K      West Toronto   
102      M6K      West Toronto   
106      M1L       Scarborough   
107      M1L       Scarborough   
108      M1L       Scarborough   
109      M2L        North York   
110      M2L        North York   
111      M3L        North York   
112      M4L      East Toronto   
113      M4L      East Toronto   
114      M5L  Downtown Toronto   
115      M5L  Downtown Toronto   
116      M6L        North York   
117      M6L        North York   
118      M6L        North York   
121      M9L        North York   
122      M1M       Scarborough   
123      M1M       Scarborough   
124      M1M       Scarborough   
125      M2M        North York   
126      M2M        North York   
127      M3M        North York   
128      M4M      East Toronto   
129      M5M        North York   
130      M5M        North York   
131      M6M              York   
132      M6M              York   
133      M6M              York   
134      M6M              York   
137      M9M        North York   
138      M9M        North York   
139      M1N       Scarborough   
140      M1N       Scarborough   
141      M2N        North York   
142      M3N        North York   
143      M4N   Central Toronto   
144      M5N   Central Toronto   
145      M6N              York   
146      M6N              York   
149      M9N              York   
150      M1P       Scarbor

If a cell has a Borough assigned but a *'Not assigned'* neighborhood, then the neighborhood will be the same as the Borough. So for the 9th cell in the table on the Wikipedia page, the value of the Borough and the Neighborhood columns will be Queen's Park.

In [18]:
df_Toronto['Neighbourhood'].mask(df_Toronto['Neighbourhood'] == "Not assigned", df_Toronto['Borough'], inplace=True)
df_Toronto

/Users/KT/opt/anaconda3/lib/python3.7/site-packages/pandas/core/generic.py:9114: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._update_inplace(new_data)


Postcode           Borough  \
2        M3A        North York   
3        M4A        North York   
4        M5A  Downtown Toronto   
5        M6A        North York   
6        M6A        North York   
7        M7A      Queen's Park   
9        M9A  Downtown Toronto   
10       M1B       Scarborough   
11       M1B       Scarborough   
13       M3B        North York   
14       M4B         East York   
15       M4B         East York   
16       M5B  Downtown Toronto   
17       M5B  Downtown Toronto   
18       M6B        North York   
21       M9B         Etobicoke   
22       M9B         Etobicoke   
23       M9B         Etobicoke   
24       M9B         Etobicoke   
25       M9B         Etobicoke   
26       M1C       Scarborough   
27       M1C       Scarborough   
28       M1C       Scarborough   
30       M3C        North York   
31       M3C        North York   
32       M4C         East York   
33       M5C  Downtown Toronto   
34       M6C              York   
37       M9C         Etobicoke   
38       M9C         Etobicoke   
39       M9C         Etobicoke   
40       M9C         Etobicoke   
41       M1E       Scarborough   
42       M1E       Scarborough   
43       M1E       Scarborough   
46       M4E      East Toronto   
47       M5E  Downtown Toronto   
48       M6E              York   
52       M1G       Scarborough   
55       M4G         East York   
56       M5G  Downtown Toronto   
57       M6G  Downtown Toronto   
61       M1H       Scarborough   
62       M2H        North York   
63       M3H        North York   
64       M3H        North York   
65       M3H        North York   
66       M4H         East York   
67       M5H  Downtown Toronto   
68       M5H  Downtown Toronto   
69       M5H  Downtown Toronto   
70       M6H      West Toronto   
71       M6H      West Toronto   
75       M1J       Scarborough   
76       M2J        North York   
77       M2J        North York   
78       M2J        North York   
79       M3J        North York   
80       M3J        North York   
81       M4J         East York   
82       M5J  Downtown Toronto   
83       M5J  Downtown Toronto   
84       M5J  Downtown Toronto   
85       M6J      West Toronto   
86       M6J      West Toronto   
90       M1K       Scarborough   
91       M1K       Scarborough   
92       M1K       Scarborough   
93       M2K        North York   
94       M3K        North York   
95       M3K        North York   
96       M4K      East Toronto   
97       M4K      East Toronto   
98       M5K  Downtown Toronto   
99       M5K  Downtown Toronto   
100      M6K      West Toronto   
101      M6K      West Toronto   
102      M6K      West Toronto   
106      M1L       Scarborough   
107      M1L       Scarborough   
108      M1L       Scarborough   
109      M2L        North York   
110      M2L        North York   
111      M3L        North York   
112      M4L      East Toronto   
113      M4L      East Toronto   
114      M5L  Downtown Toronto   
115      M5L  Downtown Toronto   
116      M6L        North York   
117      M6L        North York   
118      M6L        North York   
121      M9L        North York   
122      M1M       Scarborough   
123      M1M       Scarborough   
124      M1M       Scarborough   
125      M2M        North York   
126      M2M        North York   
127      M3M        North York   
128      M4M      East Toronto   
129      M5M        North York   
130      M5M        North York   
131      M6M              York   
132      M6M              York   
133      M6M              York   
134      M6M              York   
137      M9M        North York   
138      M9M        North York   
139      M1N       Scarborough   
140      M1N       Scarborough   
141      M2N        North York   
142      M3N        North York   
143      M4N   Central Toronto   
144      M5N   Central Toronto   
145      M6N              York   
146      M6N              York   
149      M9N              York   
150      M1P       Scarbor

More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma as shown in row 11 in the above table.

In [19]:
df_Toronto_clean = df_Toronto.groupby(by=['Postcode','Borough']).agg(lambda x: ','.join(x))

df_Toronto_clean.reset_index(level=['Postcode','Borough'], inplace=True)

In [20]:
df_Toronto_clean

Postcode           Borough  \
0        M1B       Scarborough   
1        M1C       Scarborough   
2        M1E       Scarborough   
3        M1G       Scarborough   
4        M1H       Scarborough   
5        M1J       Scarborough   
6        M1K       Scarborough   
7        M1L       Scarborough   
8        M1M       Scarborough   
9        M1N       Scarborough   
10       M1P       Scarborough   
11       M1R       Scarborough   
12       M1S       Scarborough   
13       M1T       Scarborough   
14       M1V       Scarborough   
15       M1W       Scarborough   
16       M1X       Scarborough   
17       M2H        North York   
18       M2J        North York   
19       M2K        North York   
20       M2L        North York   
21       M2M        North York   
22       M2N        North York   
23       M2P        North York   
24       M2R        North York   
25       M3A        North York   
26       M3B        North York   
27       M3C        North York   
28       M3H        North York   
29       M3J        North York   
30       M3K        North York   
31       M3L        North York   
32       M3M        North York   
33       M3N        North York   
34       M4A        North York   
35       M4B         East York   
36       M4C         East York   
37       M4E      East Toronto   
38       M4G         East York   
39       M4H         East York   
40       M4J         East York   
41       M4K      East Toronto   
42       M4L      East Toronto   
43       M4M      East Toronto   
44       M4N   Central Toronto   
45       M4P   Central Toronto   
46       M4R   Central Toronto   
47       M4S   Central Toronto   
48       M4T   Central Toronto   
49       M4V   Central Toronto   
50       M4W  Downtown Toronto   
51       M4X  Downtown Toronto   
52       M4Y  Downtown Toronto   
53       M5A  Downtown Toronto   
54       M5B  Downtown Toronto   
55       M5C  Downtown Toronto   
56       M5E  Downtown Toronto   
57       M5G  Downtown Toronto   
58       M5H  Downtown Toronto   
59       M5J  Downtown Toronto   
60       M5K  Downtown Toronto   
61       M5L  Downtown Toronto   
62       M5M        North York   
63       M5N   Central Toronto   
64       M5P   Central Toronto   
65       M5R   Central Toronto   
66       M5S  Downtown Toronto   
67       M5T  Downtown Toronto   
68       M5V  Downtown Toronto   
69       M5W  Downtown Toronto   
70       M5X  Downtown Toronto   
71       M6A        North York   
72       M6B        North York   
73       M6C              York   
74       M6E              York   
75       M6G  Downtown Toronto   
76       M6H      West Toronto   
77       M6J      West Toronto   
78       M6K      West Toronto   
79       M6L        North York   
80       M6M              York   
81       M6N              York   
82       M6P      West Toronto   
83       M6R      West Toronto   
84       M6S      West Toronto   
85       M7A      Queen's Park   
86       M7R       Mississauga   
87       M7Y      East Toronto   
88       M8V         Etobicoke   
89       M8W         Etobicoke   
90       M8X         Etobicoke   
91       M8Y         Etobicoke   
92       M8Z         Etobicoke   
93       M9A  Downtown Toronto   
94       M9B         Etobicoke   
95       M9C         Etobicoke   
96       M9L        North York   
97       M9M        North York   
98       M9N              York   
99       M9P         Etobicoke   
100      M9R         Etobicoke   
101      M9V         Etobicoke   
102      M9W         Etobicoke   

                                         Neighbourhood  
0                                        Rouge,Malvern  
1                 Highland Creek,Rouge Hill,Port Union  
2                      Guildwood,Morningside,West Hill  
3                                               Woburn  
4                                            Cedarbrae  
5                                  Scarborough Village  
6            East Birchmount Park,Ionview,Kennedy Park  
7          

In [21]:
df_Toronto_clean.shape

(103, 3)

- In the last cell of your notebook, use the .shape method to print the number of rows of your dataframe.

In [22]:
print("The number of rows in this dataset: ",df_Toronto_clean.shape[0])
print("The number of columns in this dataset: ",df_Toronto_clean.shape[1])

The number of rows in this dataset:  103
The number of columns in this dataset:  3


## 3) Get map co-ordinates: Longitude & Latitude

In order to generate co-ordinates, we would need to install and import the geocoder library.

In [23]:
!conda install -c conda-forge geocoder -y

Solving environment: done

# All requested packages already installed.



In [ ]:
import geocoder # import geocoder

# initialize your variable to None
lat_lng_coords = None

# loop until you get the coordinates
while(lat_lng_coords is None):
  g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
  lat_lng_coords = g.latlng

latitude = lat_lng_coords[0]
longitude = lat_lng_coords[1]

An alternative to getting the co-ordinates is to use the .csv file provided.

In [24]:
co_ordinates = pd.read_csv("Geospatial_Coordinates.csv")
co_ordinates.rename(columns={"Postal Code":"Postcode"}, inplace=True)
co_ordinates.head()

Postcode   Latitude  Longitude
0      M1B  43.806686 -79.194353
1      M1C  43.784535 -79.160497
2      M1E  43.763573 -79.188711
3      M1G  43.770992 -79.216917
4      M1H  43.773136 -79.239476

In [25]:
df_Toronto_clean2 = pd.merge(df_Toronto_clean, 
                             co_ordinates[['Postcode','Latitude','Longitude']], 
                             on='Postcode')
df_Toronto_clean2


Postcode           Borough  \
0        M1B       Scarborough   
1        M1C       Scarborough   
2        M1E       Scarborough   
3        M1G       Scarborough   
4        M1H       Scarborough   
5        M1J       Scarborough   
6        M1K       Scarborough   
7        M1L       Scarborough   
8        M1M       Scarborough   
9        M1N       Scarborough   
10       M1P       Scarborough   
11       M1R       Scarborough   
12       M1S       Scarborough   
13       M1T       Scarborough   
14       M1V       Scarborough   
15       M1W       Scarborough   
16       M1X       Scarborough   
17       M2H        North York   
18       M2J        North York   
19       M2K        North York   
20       M2L        North York   
21       M2M        North York   
22       M2N        North York   
23       M2P        North York   
24       M2R        North York   
25       M3A        North York   
26       M3B        North York   
27       M3C        North York   
28       M3H        North York   
29       M3J        North York   
30       M3K        North York   
31       M3L        North York   
32       M3M        North York   
33       M3N        North York   
34       M4A        North York   
35       M4B         East York   
36       M4C         East York   
37       M4E      East Toronto   
38       M4G         East York   
39       M4H         East York   
40       M4J         East York   
41       M4K      East Toronto   
42       M4L      East Toronto   
43       M4M      East Toronto   
44       M4N   Central Toronto   
45       M4P   Central Toronto   
46       M4R   Central Toronto   
47       M4S   Central Toronto   
48       M4T   Central Toronto   
49       M4V   Central Toronto   
50       M4W  Downtown Toronto   
51       M4X  Downtown Toronto   
52       M4Y  Downtown Toronto   
53       M5A  Downtown Toronto   
54       M5B  Downtown Toronto   
55       M5C  Downtown Toronto   
56       M5E  Downtown Toronto   
57       M5G  Downtown Toronto   
58       M5H  Downtown Toronto   
59       M5J  Downtown Toronto   
60       M5K  Downtown Toronto   
61       M5L  Downtown Toronto   
62       M5M        North York   
63       M5N   Central Toronto   
64       M5P   Central Toronto   
65       M5R   Central Toronto   
66       M5S  Downtown Toronto   
67       M5T  Downtown Toronto   
68       M5V  Downtown Toronto   
69       M5W  Downtown Toronto   
70       M5X  Downtown Toronto   
71       M6A        North York   
72       M6B        North York   
73       M6C              York   
74       M6E              York   
75       M6G  Downtown Toronto   
76       M6H      West Toronto   
77       M6J      West Toronto   
78       M6K      West Toronto   
79       M6L        North York   
80       M6M              York   
81       M6N              York   
82       M6P      West Toronto   
83       M6R      West Toronto   
84       M6S      West Toronto   
85       M7A      Queen's Park   
86       M7R       Mississauga   
87       M7Y      East Toronto   
88       M8V         Etobicoke   
89       M8W         Etobicoke   
90       M8X         Etobicoke   
91       M8Y         Etobicoke   
92       M8Z         Etobicoke   
93       M9A  Downtown Toronto   
94       M9B         Etobicoke   
95       M9C         Etobicoke   
96       M9L        North York   
97       M9M        North York   
98       M9N              York   
99       M9P         Etobicoke   
100      M9R         Etobicoke   
101      M9V         Etobicoke   
102      M9W         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                        Rouge,Malvern  43.806686 -79.194353  
1                 Highland Creek,Rouge Hill,Port Union  43.784535 -79.160497  
2                      Guildwood,Morningside,West Hill  43.763573 -79.188711  
3                                               Woburn  43.770992 -79.216917  
4                                            Cedarbrae  43.773136 -79.23

In [26]:
print('The dataframe has {} Boroughs and {} Neighbourhoods.'.format(
        len(df_Toronto_clean2['Borough'].unique()),
        df_Toronto_clean2.shape[0]
    )
)

The dataframe has 11 Boroughs and 103 Neighbourhoods.


#### Use geopy library to get the latitude and longitude values of Toronto, Ontario.

In [27]:
address = 'Toronto,Ontario'

In [28]:
geolocator = Nominatim(user_agent="foursquare_agent")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude

In [29]:
print("Latitude: ",latitude, "and Longitude: ", longitude)

Latitude:  43.653963 and Longitude:  -79.387207


#### Create a map of Toronto, Ontario with neighbourhoods superimposed on top

In [30]:
# create map of New York using latitude and longitude values
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, Borough, Neighbourhood in zip(df_Toronto_clean2['Latitude'], df_Toronto_clean2['Longitude'], df_Toronto_clean2['Borough'], df_Toronto_clean2['Neighbourhood']):
    label = '{}, {}'.format(Neighbourhood, Borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_Toronto)  
    
map_Toronto

**Folium** is a great visualization library. We can zoom into the above map, and click on each circle mark to reveal the name of the neighborhood and its respective borough.

downtown_toronto_data = df_Toronto_clean2[df_Toronto_clean2['Borough'] == 'Downtown Toronto'].reset_index(drop=True)
downtown_toronto_data.head()

### Define Foursquare Credentials and Version

In [31]:
CLIENT_ID = 'VAN43G4DVSK1UFQMXKX2C50SDTTRENUUPKGTNFYXR0EQ5PPR' # your Foursquare ID
CLIENT_SECRET = 'WRAK0FOE22SFDI0INIBYBO5FBSUUHO4B5KIAQW4WQ43KSNWS' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: VAN43G4DVSK1UFQMXKX2C50SDTTRENUUPKGTNFYXR0EQ5PPR
CLIENT_SECRET:WRAK0FOE22SFDI0INIBYBO5FBSUUHO4B5KIAQW4WQ43KSNWS


Get the neighborhood's latitude and longitude values. Let's test the first entry.

In [32]:
neighbourhood_latitude = df_Toronto_clean2.loc[0, 'Latitude'] # neighborhood latitude value
neighbourhood_longitude = df_Toronto_clean2.loc[0, 'Longitude'] # neighborhood longitude value

neighbourhood_name = df_Toronto_clean2.loc[0, 'Neighbourhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighbourhood_name, 
                                                               neighbourhood_latitude, 
                                                               neighbourhood_longitude))

Latitude and longitude values of Rouge,Malvern are 43.806686299999996, -79.19435340000001.


Now, let's get the top 100 venues that are in Rouge,Malvern within a radius of 500 meters.
First, let's create the GET request URL. Name your URL url.

In [33]:
# type your answer here

LIMIT = 100 # limit of number of venues returned by Foursquare API

radius = 500 # define radius

# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighbourhood_latitude, 
    neighbourhood_longitude, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=VAN43G4DVSK1UFQMXKX2C50SDTTRENUUPKGTNFYXR0EQ5PPR&client_secret=WRAK0FOE22SFDI0INIBYBO5FBSUUHO4B5KIAQW4WQ43KSNWS&v=20180605&ll=43.806686299999996,-79.19435340000001&radius=500&limit=100'

Send the GET request and examine the results.

In [34]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5df1b568f7706a001b48ea4f'},
 'response': {'warning': {'text': "There aren't a lot of results near you. Try something more general, reset your filters, or expand the search area."},
  'headerLocation': 'Malvern',
  'headerFullLocation': 'Malvern, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 2,
  'suggestedBounds': {'ne': {'lat': 43.8111863045, 'lng': -79.18812958073042},
   'sw': {'lat': 43.80218629549999, 'lng': -79.2005772192696}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4bb6b9446edc76b0d771311c',
       'name': "Wendy's",
       'location': {'crossStreet': 'Morningside & Sheppard',
        'lat': 43.80744841934756,
        'lng': -79.19905558052072,
        'labeledLatLngs': [{'label': 'display'

Let's research what is around these destinations.

In [35]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Now we are ready to clean the json and structure it into a pandas dataframe.

In [36]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name            categories        lat        lng
0                Wendy's  Fast Food Restaurant  43.807448 -79.199056
1  Interprovincial Group            Print Shop  43.805630 -79.200378

In [37]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

2 venues were returned by Foursquare.


## 4) Explore Neighborhoods in Toronto, Ontario

Let's create a function to repeat the same process to all the neighborhoods in Toronto, Ontario.

In [38]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Now write the code to run the above function on each neighborhood and create a new dataframe called Toronto_venues.

In [39]:
Toronto_venues = getNearbyVenues(names=df_Toronto_clean2['Neighbourhood'],
                                   latitudes=df_Toronto_clean2['Latitude'],
                                   longitudes=df_Toronto_clean2['Longitude']
                                  )

Rouge,Malvern
Highland Creek,Rouge Hill,Port Union
Guildwood,Morningside,West Hill
Woburn
Cedarbrae
Scarborough Village
East Birchmount Park,Ionview,Kennedy Park
Clairlea,Golden Mile,Oakridge
Cliffcrest,Cliffside,Scarborough Village West
Birch Cliff,Cliffside West
Dorset Park,Scarborough Town Centre,Wexford Heights
Maryvale,Wexford
Agincourt
Clarks Corners,Sullivan,Tam O'Shanter
Agincourt North,L'Amoreaux East,Milliken,Steeles East
L'Amoreaux West
Upper Rouge
Hillcrest Village
Fairview,Henry Farm,Oriole
Bayview Village
Silver Hills,York Mills
Newtonbrook,Willowdale
Willowdale South
York Mills West
Willowdale West
Parkwoods
Don Mills North
Flemingdon Park,Don Mills South
Bathurst Manor,Downsview North,Wilson Heights
Northwood Park,York University
CFB Toronto,Downsview East
Downsview West
Downsview Central
Downsview Northwest
Victoria Village
Woodbine Gardens,Parkview Hill
Woodbine Heights
The Beaches
Leaside
Thorncliffe Park
East Toronto
The Danforth West,Riverdale
The Beaches West,Indi

#### Let's check the size of the resulting dataframe

In [40]:
print(Toronto_venues.shape)
Toronto_venues.head()

(2241, 7)


Neighbourhood  Neighbourhood Latitude  \
0                         Rouge,Malvern               43.806686   
1                         Rouge,Malvern               43.806686   
2  Highland Creek,Rouge Hill,Port Union               43.784535   
3  Highland Creek,Rouge Hill,Port Union               43.784535   
4  Highland Creek,Rouge Hill,Port Union               43.784535   

   Neighbourhood Longitude                      Venue  Venue Latitude  \
0               -79.194353                    Wendy's       43.807448   
1               -79.194353      Interprovincial Group       43.805630   
2               -79.160497     Chris Effects Painting       43.784343   
3               -79.160497      Royal Canadian Legion       43.782533   
4               -79.160497  Affordable Toronto Movers       43.787919   

   Venue Longitude              Venue Category  
0       -79.199056        Fast Food Restaurant  
1       -79.200378                  Print Shop  
2       -79.163742  Construction & Landscaping  
3       -79.163085                         Bar  
4       -79.162977               Moving Target

Let's check how many venues were returned for each neighborhood

In [41]:
Toronto_venues.groupby('Neighbourhood').count()

Neighbourhood Latitude  \
Neighbourhood                                                                
Adelaide,King,Richmond                                                 100   
Agincourt                                                                5   
Agincourt North,L'Amoreaux East,Milliken,Steele...                       3   
Albion Gardens,Beaumond Heights,Humbergate,Jame...                       8   
Alderwood,Long Branch                                                    9   
Bathurst Manor,Downsview North,Wilson Heights                           19   
Bayview Village                                                          4   
Bedford Park,Lawrence Manor East                                        23   
Berczy Park                                                             55   
Birch Cliff,Cliffside West                                               4   
Bloordale Gardens,Eringate,Markland Wood,Old Bu...                       7   
Brockton,Exhibition Place,Parkdale Village                              22   
Business Reply Mail Processing Centre 969 Eastern                       16   
CFB Toronto,Downsview East                                               2   
CN Tower,Bathurst Quay,Island airport,Harbourfr...                      15   
Cabbagetown,St. James Town                                              43   
Caledonia-Fairbanks                                                      5   
Canada Post Gateway Processing Centre                                   10   
Cedarbrae                                                                8   
Central Bay Street                                                      84   
Chinatown,Grange Park,Kensington Market                                 94   
Christie                                                                17   
Church and Wellesley                                                    83   
Clairlea,Golden Mile,Oakridge                                           10   
Clarks Corners,Sullivan,Tam O'Shanter                                   12   
Cliffcrest,Cliffside,Scarborough Village West                            2   
Cloverdale,Islington,Martin Grove,Princess Gard...                       2   
Commerce Court,Victoria Hotel                                          100   
Davisville                                                              35   
Davisville North                                                         9   
Deer Park,Forest Hill SE,Rathnelly,South Hill,S...                      15   
Del Ray,Keelesdale,Mount Dennis,Silverthorn                              4   
Design Exchange,Toronto Dominion Centre                                100   
Don Mills North                                                          4   
Dorset Park,Scarborough Town Centre,Wexford Hei...                       7   
Dovercourt Village,Dufferin                                             17   
Downsview Central                                                        3   
Downsview Northwest                                                      4   
Downsview West                                                           6   
Downsview,North Park,Upwood Park                                         3   
East Birchmount Park,Ionview,Kennedy Park                                8   
East Toronto                                                             4   
Emery,Humberlea                                                          2   
Fairview,Henry Farm,Oriole                                              66   
First Canadian Place,Underground city                                  100   
Flemingdon Park,Don Mills South                                         21   
Forest Hill North,Forest Hill West                                       4   
Glencairn                                                                5   
Guildwood,Morningside,West Hill                                          9   
Harbord,University of Toronto                                           38

#### Let's find out how many unique categories can be curated from all the returned venues

In [42]:
print('There are {} uniques categories.'.format(len(Toronto_venues['Venue Category'].unique())))

There are 272 uniques categories.


## 5) Analyze each Neighbourhood

In [44]:
# one hot encoding
Toronto_onehot = pd.get_dummies(Toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
Toronto_onehot['Neighbourhood'] = Toronto_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [Toronto_onehot.columns[-1]] + list(Toronto_onehot.columns[:-1])
Toronto_onehot = Toronto_onehot[fixed_columns]

Toronto_onehot.head()

Neighbourhood  Accessories Store  Afghan Restaurant  \
0                         Rouge,Malvern                  0                  0   
1                         Rouge,Malvern                  0                  0   
2  Highland Creek,Rouge Hill,Port Union                  0                  0   
3  Highland Creek,Rouge Hill,Port Union                  0                  0   
4  Highland Creek,Rouge Hill,Port Union                  0                  0   

   Airport  Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0        0                   0             0               0                0   
1        0                   0             0               0                0   
2        0                   0             0               0                0   
3        0                   0             0               0                0   
4        0                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  \
0                 0                    0             0         0   
1                 0                    0             0         0   
2                 0                    0             0         0   
3                 0                    0             0         0   
4                 0                    0             0         0   

   Argentinian Restaurant  Art Gallery  Arts & Crafts Store  Asian Restaurant  \
0                       0            0                    0                 0   
1                       0            0                    0                 0   
2                       0            0                    0                 0   
3                       0            0                    0                 0   
4                       0            0                    0                 0   

   Athletics & Sports  Auto Garage  Auto Workshop  BBQ Joint  Baby Store  \
0                   0            0              0          0           0   
1                   0            0              0          0           0   
2                   0            0              0          0           0   
3                   0            0              0          0           0   
4                   0            0              0          0           0   

   Bagel Shop  Bakery  Bank  Bar  Baseball Field  Baseball Stadium  \
0           0       0     0    0               0                 0   
1           0       0     0    0               0                 0   
2           0       0     0    0               0                 0   
3           0       0     0    1               0                 0   
4           0       0     0    0               0                 0   

   Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  Beer Store  \
0                   0      0                0         0           0   
1                   0      0                0         0           0   
2                   0      0                0         0           0   
3                   0      0                0         0           0   
4                   0      0                0         0           0   

   Belgian Restaurant  Bike Shop  Bistro  Board Shop  Boat or Ferry  \
0                   0          0       0           0              0   
1                   0          0       0           0              0   
2                   0          0       0           0              0   
3                   0          0       0           0              0   
4                   0          0       0           0              0   

   Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  Brewery  \
0          0         0                     0               0        0   
1          0         0                     0               0        0   
2          0         0                     0               0        0   
3          0         0                     0               0        0   
4          0         0                     0               

In [45]:
Toronto_onehot.shape

(2241, 273)

Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [46]:
Toronto_grouped = Toronto_onehot.groupby('Neighbourhood').mean().reset_index()
Toronto_grouped

Neighbourhood  Accessories Store  \
0                               Adelaide,King,Richmond           0.000000   
1                                            Agincourt           0.000000   
2    Agincourt North,L'Amoreaux East,Milliken,Steel...           0.000000   
3    Albion Gardens,Beaumond Heights,Humbergate,Jam...           0.000000   
4                                Alderwood,Long Branch           0.000000   
5        Bathurst Manor,Downsview North,Wilson Heights           0.000000   
6                                      Bayview Village           0.000000   
7                     Bedford Park,Lawrence Manor East           0.000000   
8                                          Berczy Park           0.000000   
9                           Birch Cliff,Cliffside West           0.000000   
10   Bloordale Gardens,Eringate,Markland Wood,Old B...           0.000000   
11          Brockton,Exhibition Place,Parkdale Village           0.000000   
12   Business Reply Mail Processing Centre 969 Eastern           0.000000   
13                          CFB Toronto,Downsview East           0.000000   
14   CN Tower,Bathurst Quay,Island airport,Harbourf...           0.000000   
15                          Cabbagetown,St. James Town           0.000000   
16                                 Caledonia-Fairbanks           0.000000   
17               Canada Post Gateway Processing Centre           0.000000   
18                                           Cedarbrae           0.000000   
19                                  Central Bay Street           0.000000   
20             Chinatown,Grange Park,Kensington Market           0.000000   
21                                            Christie           0.000000   
22                                Church and Wellesley           0.000000   
23                       Clairlea,Golden Mile,Oakridge           0.000000   
24               Clarks Corners,Sullivan,Tam O'Shanter           0.000000   
25       Cliffcrest,Cliffside,Scarborough Village West           0.000000   
26   Cloverdale,Islington,Martin Grove,Princess Gar...           0.000000   
27                       Commerce Court,Victoria Hotel           0.000000   
28                                          Davisville           0.000000   
29                                    Davisville North           0.000000   
30   Deer Park,Forest Hill SE,Rathnelly,South Hill,...           0.000000   
31         Del Ray,Keelesdale,Mount Dennis,Silverthorn           0.000000   
32             Design Exchange,Toronto Dominion Centre           0.000000   
33                                     Don Mills North           0.000000   
34   Dorset Park,Scarborough Town Centre,Wexford He...           0.000000   
35                         Dovercourt Village,Dufferin           0.000000   
36                                   Downsview Central           0.000000   
37                                 Downsview Northwest           0.000000   
38                                      Downsview West           0.000000   
39                    Downsview,North Park,Upwood Park           0.000000   
40           East Birchmount Park,Ionview,Kennedy Park           0.000000   
41                                        East Toronto           0.000000   
42                                     Emery,Humberlea           0.000000   
43                          Fairview,Henry Farm,Oriole           0.015152   
44               First Canadian Place,Underground city           0.000000   
45                     Flemingdon Park,Don Mills South           0.000000   
46                  Forest Hill North,Forest Hill West           0.000000   
47                                           Glencairn           0.000000   
48                     Guildwood,Morningside,West Hill           0.000000   
49                       Harbord,University of Toronto           0.000000   
50                                        Harbourfront           0.000000   
51     Harbourfront East,Toronto Isla

##### Let's confirm the new size

In [47]:
Toronto_grouped.shape

(101, 273)

Let's print each neighborhood along with the top 5 most common venues

In [48]:
num_top_venues = 5

for hood in Toronto_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = Toronto_grouped[Toronto_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide,King,Richmond----
             venue  freq
0      Coffee Shop  0.07
1             Café  0.05
2       Steakhouse  0.04
3  Thai Restaurant  0.04
4           Bakery  0.03


----Agincourt----
                       venue  freq
0                     Lounge   0.2
1             Breakfast Spot   0.2
2  Latin American Restaurant   0.2
3               Skating Rink   0.2
4             Clothing Store   0.2


----Agincourt North,L'Amoreaux East,Milliken,Steeles East----
                 venue  freq
0                 Park  0.33
1           Playground  0.33
2  Arts & Crafts Store  0.33
3    Accessories Store  0.00
4   Mexican Restaurant  0.00


----Albion Gardens,Beaumond Heights,Humbergate,Jamestown,Mount Olive,Silverstone,South Steeles,Thistletown----
                  venue  freq
0   Fried Chicken Joint  0.12
1         Grocery Store  0.12
2              Pharmacy  0.12
3  Fast Food Restaurant  0.12
4           Pizza Place  0.12


----Alderwood,Long Branch----
          venue  freq
0   

4  Accessories Store  0.00


----Glencairn----
                 venue  freq
0                  Pub   0.2
1                 Park   0.2
2     Asian Restaurant   0.2
3        Metro Station   0.2
4  Japanese Restaurant   0.2


----Guildwood,Morningside,West Hill----
                venue  freq
0   Electronics Store  0.22
1  Mexican Restaurant  0.11
2                 Spa  0.11
3         Pizza Place  0.11
4      Breakfast Spot  0.11


----Harbord,University of Toronto----
                 venue  freq
0                 Café  0.16
1           Restaurant  0.05
2   Italian Restaurant  0.05
3  Japanese Restaurant  0.05
4               Bakery  0.05


----Harbourfront----
                venue  freq
0         Coffee Shop  0.14
1                Park  0.08
2                 Pub  0.06
3              Bakery  0.06
4  Mexican Restaurant  0.04


----Harbourfront East,Toronto Islands,Union Station----
            venue  freq
0     Coffee Shop  0.13
1        Aquarium  0.05
2           Hotel  0.05
3         

4           Bookstore  0.05


----The Junction North,Runnymede----
                  venue  freq
0  Caribbean Restaurant  0.25
1         Grocery Store  0.25
2               Brewery  0.25
3              Bus Line  0.25
4     Accessories Store  0.00


----The Kingsway,Montgomery Road,Old Mill North----
                             venue  freq
0                             Pool   0.5
1                            River   0.5
2                    Metro Station   0.0
3  Molecular Gastronomy Restaurant   0.0
4       Modern European Restaurant   0.0


----Thorncliffe Park----
               venue  freq
0  Indian Restaurant  0.11
1       Burger Joint  0.11
2                Gym  0.06
3               Bank  0.06
4    Warehouse Store  0.06


----Victoria Village----
                             venue  freq
0                     Hockey Arena  0.25
1            Portuguese Restaurant  0.25
2                French Restaurant  0.25
3                      Coffee Shop  0.25
4  Molecular Gastronomy Restaura

Let's put that into a pandas dataframe

First, let's write a function to sort the venues in descending order.

In [49]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [60]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = Toronto_grouped['Neighbourhood']

for ind in np.arange(Toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(Toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood  \
0                             Adelaide,King,Richmond   
1                                          Agincourt   
2  Agincourt North,L'Amoreaux East,Milliken,Steel...   
3  Albion Gardens,Beaumond Heights,Humbergate,Jam...   
4                              Alderwood,Long Branch   

       1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0                Coffee Shop                  Café       Thai Restaurant   
1  Latin American Restaurant          Skating Rink                Lounge   
2                       Park   Arts & Crafts Store            Playground   
3              Grocery Store   Fried Chicken Joint  Fast Food Restaurant   
4                Pizza Place          Skating Rink           Coffee Shop   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0            Steakhouse                   Bar                Bakery   
1        Clothing Store        Breakfast Spot            Donut Shop   
2      Doner Restaurant          Dessert Shop    Dim Sum Restaurant   
3              Pharmacy           Video Store            Beer Store   
4                   Gym                   Pub              Pharmacy   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0            Restaurant           Salad Place      Sushi Restaurant   
1    Dim Sum Restaurant                 Diner        Discount Store   
2                 Diner        Discount Store               Dog Run   
3           Pizza Place        Sandwich Place               Dog Run   
4          Dance Studio        Sandwich Place        Discount Store   

  10th Most Common Venue  
0       Asian Restaurant  
1                Dog Run  
2             Donut Shop  
3           Dessert Shop  
4       Department Store

## 6) Cluster Neighborhoods

Run k-means to cluster the neighborhood into 5 clusters.

In [61]:
# set number of clusters
kclusters = 5

Toronto_grouped_clustering = Toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(Toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([4, 4, 2, 1, 1, 1, 4, 1, 4, 4], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [62]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Toronto_merged = df_Toronto_clean2

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
Toronto_merged = Toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), 
                                     on='Neighbourhood', 
                                     how = 'right')

Toronto_merged.head() # check the last columns!

Postcode      Borough                         Neighbourhood   Latitude  \
0      M1B  Scarborough                         Rouge,Malvern  43.806686   
1      M1C  Scarborough  Highland Creek,Rouge Hill,Port Union  43.784535   
2      M1E  Scarborough       Guildwood,Morningside,West Hill  43.763573   
3      M1G  Scarborough                                Woburn  43.770992   
4      M1H  Scarborough                             Cedarbrae  43.773136   

   Longitude  Cluster Labels 1st Most Common Venue  \
0 -79.194353               1  Fast Food Restaurant   
1 -79.160497               4         Moving Target   
2 -79.188711               4     Electronics Store   
3 -79.216917               1           Coffee Shop   
4 -79.239476               4           Gas Station   

        2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Print Shop           Yoga Studio      Doner Restaurant   
1  Construction & Landscaping                   Bar           Yoga Studio   
2              Medical Center    Mexican Restaurant                   Spa   
3                    Pharmacy     Korean Restaurant           Yoga Studio   
4                      Bakery   Fried Chicken Joint  Caribbean Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0          Dessert Shop    Dim Sum Restaurant                 Diner   
1      Doner Restaurant    Dim Sum Restaurant                 Diner   
2        Breakfast Spot           Pizza Place   Rental Car Location   
3            Donut Shop    Dim Sum Restaurant                 Diner   
4    Athletics & Sports       Thai Restaurant                  Bank   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0        Discount Store               Dog Run             Donut Shop  
1        Discount Store               Dog Run             Donut Shop  
2          Intersection    Dim Sum Restaurant                  Diner  
3        Discount Store               Dog Run       Doner Restaurant  
4      Hakka Restaurant   Dumpling Restaurant              Drugstore

In [63]:
neighborhoods_venues_sorted.shape

(101, 12)

In [64]:
df_Toronto_clean2.shape

(103, 5)

Finally, let's visualize the resulting clusters

In [65]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(Toronto_merged['Latitude'], Toronto_merged['Longitude'], Toronto_merged['Neighbourhood'], Toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## 7) Examine Clusters

Now, you can examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, you can then assign a name to each cluster. I will leave this exercise to you.

### Cluster 1

In [66]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 0, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
20  North York               0             Cafeteria            Donut Shop   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
20          Dessert Shop    Dim Sum Restaurant                 Diner   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
20        Discount Store               Dog Run      Doner Restaurant   

   9th Most Common Venue 10th Most Common Venue  
20             Drugstore     College Rec Center

### Cluster 2

In [67]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 1, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels     1st Most Common Venue  \
0        Scarborough               1      Fast Food Restaurant   
3        Scarborough               1               Coffee Shop   
8        Scarborough               1       American Restaurant   
13       Scarborough               1               Pizza Place   
15       Scarborough               1      Fast Food Restaurant   
17        North York               1  Mediterranean Restaurant   
24        North York               1                   Butcher   
28        North York               1               Coffee Shop   
33        North York               1             Grocery Store   
35         East York               1      Fast Food Restaurant   
39         East York               1              Burger Joint   
47   Central Toronto               1            Sandwich Place   
49   Central Toronto               1               Coffee Shop   
62        North York               1          Sushi Restaurant   
65   Central Toronto               1            Sandwich Place   
80              York               1         Convenience Store   
86       Mississauga               1                     Hotel   
88         Etobicoke               1               Coffee Shop   
89         Etobicoke               1               Pizza Place   
95         Etobicoke               1         Convenience Store   
96        North York               1       Empanada Restaurant   
99         Etobicoke               1               Pizza Place   
100        Etobicoke               1                      Park   
101        Etobicoke               1             Grocery Store   

    2nd Most Common Venue      3rd Most Common Venue  \
0              Print Shop                Yoga Studio   
3                Pharmacy          Korean Restaurant   
8                   Motel              Deli / Bodega   
13          Shopping Mall       Fast Food Restaurant   
15     Chinese Restaurant              Grocery Store   
17            Golf Course       Fast Food Restaurant   
24            Coffee Shop                   Pharmacy   
28          Shopping Mall                Pizza Place   
33           Liquor Store       Gym / Fitness Center   
35            Pizza Place         Athletics & Sports   
39      Indian Restaurant                Yoga Studio   
47           Dessert Shop                Pizza Place   
49                    Pub                 Bagel Shop   
62            Coffee Shop         Italian Restaurant   
65                   Café                Coffee Shop   
80            Coffee Shop             Sandwich Place   
86            Coffee Shop              Burrito Place   
88            Pizza Place             Sandwich Place   
89           Skating Rink                Coffee Shop   
95               Pharmacy                 Beer Store   
96            Pizza Place               Home Service   
99     Chinese Restaurant  Middle Eastern Restaurant   
100           Pizza Place                   Bus Line   
101   Fried Chicken Joint       Fast Food Restaurant   

        4th Most Common Venue        5th Most Common Venue  \
0            Doner Restaurant                 Dessert Shop   
3                 Yoga Studio                   Donut Shop   
8                Dessert Shop           Dim Sum Restaurant   
13        Fried Chicken Joint              Thai Restaurant   
15     Thrift / Vintage Store                          Gym   
17         Athletics & Sports                      Dog Run   
24                Pizza Place               Discount Store   
28                Supermarket                Deli / Bodega   
33         Athletics & Sports                   Donut Shop   
35               Intersection                     Bus Line   
39                Supermarket                         Bank   
47                       Café                  Coffee Shop   
49                 Sports Bar             Sushi Restaurant   
62              Grocery Store                      Butcher   
65                       Park  

Cluster 2 is more wide-spread and would require further investigation. Perhaps increasing the cluster size or extracting the population of these neighbourhoods and the salaries/wages would give us a better idea of what type of cluster this represents.

### Cluster 3

In [68]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 2, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
14       Scarborough               2                  Park   
21        North York               2                  Park   
23        North York               2                  Park   
25        North York               2                  Park   
30        North York               2                  Park   
40         East York               2                  Park   
50  Downtown Toronto               2                  Park   
74              York               2                  Park   
79        North York               2                  Park   
98              York               2     Convenience Store   

   2nd Most Common Venue       3rd Most Common Venue 4th Most Common Venue  \
14   Arts & Crafts Store                  Playground      Doner Restaurant   
21             Piano Bar            Doner Restaurant      Department Store   
23                  Bank           Convenience Store   Empanada Restaurant   
25     Food & Drink Shop                 Yoga Studio      Doner Restaurant   
30               Airport                  Donut Shop          Dessert Shop   
40           Coffee Shop           Convenience Store  Ethiopian Restaurant   
50                 Trail                  Playground      Department Store   
74  Fast Food Restaurant                      Market         Women's Store   
79                Bakery  Construction & Landscaping            Donut Shop   
98                  Park                  Donut Shop          Dessert Shop   

   5th Most Common Venue        6th Most Common Venue  \
14          Dessert Shop           Dim Sum Restaurant   
21          Dessert Shop           Dim Sum Restaurant   
23     Electronics Store  Eastern European Restaurant   
25          Dessert Shop           Dim Sum Restaurant   
30    Dim Sum Restaurant                        Diner   
40   Empanada Restaurant            Electronics Store   
50          Dessert Shop           Dim Sum Restaurant   
74             Gift Shop            German Restaurant   
79    Dim Sum Restaurant                        Diner   
98    Dim Sum Restaurant                        Diner   

          7th Most Common Venue        8th Most Common Venue  \
14                        Diner               Discount Store   
21                        Diner               Discount Store   
23          Dumpling Restaurant                    Drugstore   
25                        Diner               Discount Store   
30               Discount Store                      Dog Run   
40  Eastern European Restaurant          Dumpling Restaurant   
50                        Diner               Discount Store   
74            Electronics Store  Eastern European Restaurant   
79               Discount Store                      Dog Run   
98               Discount Store                      Dog Run   

   9th Most Common Venue 10th Most Common Venue  
14               Dog Run             Donut Shop  
21               Dog Run             Donut Shop  
23      Department Store             Donut Shop  
25               Dog Run             Donut Shop  
30      Doner Restaurant              Drugstore  
40             Drugstore          Deli / Bodega  
50               Dog Run       Doner Restaurant  
74           Golf Course    Dumpling Restaurant  
79      Doner Restaurant              Drugstore  
98      Doner Restaurant            Yoga Studio

Cluster 3 is most popular for the Park.

### Cluster 4

In [69]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 3, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
5  Scarborough               3            Playground           Yoga Studio   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
5            Donut Shop          Dessert Shop    Dim Sum Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
5                 Diner        Discount Store               Dog Run   

  9th Most Common Venue 10th Most Common Venue  
5      Doner Restaurant              Drugstore

Since this is a small cluster and could possibly be linked to cluster 3 (mostly Park).

### Cluster 5

In [70]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 4, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels      1st Most Common Venue  \
1         Scarborough               4              Moving Target   
2         Scarborough               4          Electronics Store   
4         Scarborough               4                Gas Station   
6         Scarborough               4             Discount Store   
7         Scarborough               4                     Bakery   
9         Scarborough               4                       Café   
10        Scarborough               4          Indian Restaurant   
11        Scarborough               4             Sandwich Place   
12        Scarborough               4  Latin American Restaurant   
18         North York               4             Clothing Store   
19         North York               4                       Café   
22         North York               4                Coffee Shop   
26         North York               4                       Café   
27         North York               4                Coffee Shop   
29         North York               4     Furniture / Home Store   
31         North York               4              Grocery Store   
32         North York               4           Business Service   
34         North York               4                Coffee Shop   
36          East York               4                       Park   
37       East Toronto               4       Other Great Outdoors   
38          East York               4                Coffee Shop   
41       East Toronto               4           Greek Restaurant   
42       East Toronto               4             Sandwich Place   
43       East Toronto               4                       Café   
44    Central Toronto               4                       Park   
45    Central Toronto               4                      Hotel   
46    Central Toronto               4             Clothing Store   
48    Central Toronto               4                 Playground   
51   Downtown Toronto               4                Coffee Shop   
52   Downtown Toronto               4           Sushi Restaurant   
53   Downtown Toronto               4                Coffee Shop   
54   Downtown Toronto               4                Coffee Shop   
55   Downtown Toronto               4                       Café   
56   Downtown Toronto               4                Coffee Shop   
57   Downtown Toronto               4                Coffee Shop   
58   Downtown Toronto               4                Coffee Shop   
59   Downtown Toronto               4                Coffee Shop   
60   Downtown Toronto               4                Coffee Shop   
61   Downtown Toronto               4                Coffee Shop   
63    Central Toronto               4                     Garden   
64    Central Toronto               4                       Park   
66   Downtown Toronto               4                       Café   
67   Downtown Toronto               4                       Café   
68   Downtown Toronto               4           Airport Terminal   
69   Downtown Toronto               4                Coffee Shop   
70   Downtown Toronto               4                Coffee Shop   
71         North York               4     Furniture / Home Store   
72         North York               4                       Park   
73               York               4                      Trail   
75   Downtown Toronto               4              Grocery Store   
76       West Toronto               4                   Pharmacy   
77       West Toronto               4                        Bar   
78       West Toronto               4             Breakfast Spot   
81               York               4              Grocery Store   
82       West Toronto               4         Mexican Restaurant   
83       West Toronto               4                  Gift Shop   
84       West Toronto               4                       Café   
85       Queen's Park               4                Coffee Shop   
93

Similar to cluster 2, this cluster is also very wide-spread. We would need more information to split this further. Perhaps increasing the cluster size or extracting the population of these neighbourhoods and the salaries/wages would give us a better idea of what type of cluster this represents.

#### References: 

1) https://pythonprogramminglanguage.com/web-scraping-with-pandas-and-beautifulsoup/  
2) https://buildmedia.readthedocs.org/media/pdf/geocoder/latest/geocoder.pdf  
3) https://www.shanelynn.ie/batch-geocoding-in-python-with-google-geocoding-api/  